syntax = "proto3";

package metatrader;

service DealsService {

    // Gets an array of deals.
    rpc GetDeals (GetDealsReq) returns (DealsList);
}

// https://www.mql5.com/en/docs/constants/tradingconstants/dealproperties#enum_deal_type
enum DealType {
    BUY = 0;
    SELL = 1;
    BALANCE = 2;
    CREDIT = 3;
    CHARGE = 4;
    CORRECTION = 5;
    BONUS = 6;
    COMMISSION = 7;
    COMMISSION_DAILY = 8;
    COMMISSION_MONTHLY = 9;
    COMMISSION_AGENT_DAILY = 10;
    COMMISSION_AGENT_MONTHLY = 11;
    INTEREST = 12;
    BUY_CANCELED = 13;
    SELL_CANCELED = 14;
    DDIVIDEND = 15;
    DIVIDEND_FRANKED = 16;
    TAX = 17;
}

// https://www.mql5.com/en/docs/constants/tradingconstants/dealproperties#enum_deal_reason
enum DealReason {
    CLIENT = 0;
    MOBILE = 1;
    WEB = 2;
    EXPERT = 3;
    SL = 4;
    TP = 5;
    SO = 6;
    ROLLOVER = 7;
    VMARGIN = 8;
    SPLIT = 9;
}

message Deal {
    int64 id = 1;
    string symbol = 2;
    DealType type = 3;
    float price = 4;
    float volume = 7;
    float profit = 8;
    float swap = 9;
    float fee = 10;
    int64 time = 11;
    int64 magic = 12;
    string comment = 13;
    DealReason reason = 14;
}

message GetDealsReq {
    int64 fromDate = 1;
    int64 toDate = 2;
}

message DealsList {
    repeated Deal deals = 1;
}
